File: Reqnroll/Chapter19/19.4-EmbeddedApplicationLogPattern/After/WIMP.Specs/Drivers/OrderingApiDriver.cs
using System.Net;

using Microsoft.Extensions.Logging;

using WIMP.App.Models;
using WIMP.App.RestApi;
using WIMP.Specs.Support;

namespace WIMP.Specs.Drivers;

public class OrderingApiDriver(ILoggerFactory loggerFactory, RestApiContext restApiContext)
{
    public TestAction<Order> PlaceOrder(PlaceOrderRequest placeOrderRequest) =>
        new LambdaAction<Order>(loggerFactory, "Place order", placeOrderRequest, async () =>
            await restApiContext.ProcessRequest<Order>(
                "Place order", HttpMethod.Post, "/api/orders",
                placeOrderRequest, HttpStatusCode.Created));
}

File: CucumberJVM/Chapter15/15.7-StubDependencyPattern/Before/src/test/java/com/wimp/app/specs/drivers/OrderingApiDriver.java
package com.wimp.app.specs.drivers;

import com.wimp.app.models.Order;
import com.wimp.app.restapi.PlaceOrderRequest;
import com.wimp.app.specs.support.LambdaAction;
import com.wimp.app.specs.support.TestAction;
import org.springframework.stereotype.Component;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.service.annotation.HttpExchange;
import org.springframework.web.service.annotation.PostExchange;

@Component
public class OrderingApiDriver {
    private final OrderingApiClient orderingApiClient;

    @HttpExchange("/api/orders")
    public interface OrderingApiClient {
        @PostExchange
        Order placeOrder(@RequestBody PlaceOrderRequest request);
    }

    @SuppressWarnings("SpringJavaInjectionPointsAutowiringInspection")
    public OrderingApiDriver(OrderingApiClient orderingApiClient) {
        this.orderingApiClient = orderingApiClient;
    }

    public TestAction<Order> placeOrder(PlaceOrderRequest placeOrderRequest) {
        return new LambdaAction<>("Place order", placeOrderRequest, () -> orderingApiClient.placeOrder(placeOrderRequest));
    }
}
